src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -245,7 +250,18 @@ object RefArrayUtils {
       Array.copy(array, 0, newArray, 0, end)
       newArray
     }
-    if (n <= arrayLength) array else growArray
+    def shrinkArray = {
+      var newSize: Long = math.max(arrayLength * 2, 8)
+      while (newSize > 2*n && newSize > 8)
+        newSize = newSize / 2
+      val newArray: Array[AnyRef] = new Array(newSize.toInt)
+      Array.copy(array, 0, newArray, 0, end)
+      newArray
+    }
+
+    if (n > arrayLength) growArray
+    else if (3*n <= arrayLength) shrinkArray
```

Comment:
Is the Integer overflow case covered here?

Reply:
No. Good catch!